modules/markup/html.go

```diff
@@ -634,18 +634,27 @@ func issueIndexPatternProcessor(ctx *postProcessCtx, node *html.Node) {
 		ref   *references.RenderizableReference
 	)
 
-	if ctx.metas["style"] == IssueNameStyleAlphanumeric {
-		found, ref = references.FindRenderizableReferenceAlphanumeric(node.Data)
-	} else {
-		found, ref = references.FindRenderizableReferenceNumeric(node.Data)
+	_, exttrack := ctx.metas["format"]
+	alphanum := ctx.metas["style"] == IssueNameStyleAlphanumeric
+
+	// Repos with external issue trackers might still need to reference local PRs
+	// We need to concern with the first one that shows up in the text, whichever it is
+	found, ref = references.FindRenderizableReferenceNumeric(node.Data, exttrack && alphanum)
+	if exttrack && alphanum {
+		if found2, ref2 := references.FindRenderizableReferenceAlphanumeric(node.Data); found2 {
+			if !found || ref2.RefLocation.Start < ref.RefLocation.Start {
+				found = true
+				ref = ref2
+			}
+		}
 	}
 	if !found {
 		return
 	}
 
 	var link *html.Node
 	reftext := node.Data[ref.RefLocation.Start:ref.RefLocation.End]
-	if _, ok := ctx.metas["format"]; ok {
+	if exttrack && !ref.Local {
```

Comment:
Semantically, I would say that # can be used to reference issues and pulls (except for external trackers) while ! is used to specifically reference PRs. Hence I think the link shall be to /org/repo/**pulls**/[number]. I know it will redirect in any case but it would be more clear when hovering the link.

Reply:
That is very difficult to achieve because the markdown processor does not access the database at the moment of rendering for performance reasons, so it can't know whether "1234" is a PR or an Issue. That's why I thought it would be better to make `!` mean "always local" rather than meaning PR especifically. It just happens that the local part of that scenario is the PR.

Besides, the point is to be backward compatible. If everybody likes `#1234` and use Gitea for tracking issues, I don't see any problem keeping that syntax.

Reply:
I don't mean you should validate if the referenced issue actually is a PR. Only when calling `createLink` (see some lines furhter down the code) you should use "pulls" instead of "issues. If pressing the link Gitea will redirect both ways pulls<->issue if it is of the other type.

Reply:
So if you have set
`reftype = "issues" OR "pulls `
below code will look like:
```
	var link *html.Node
	reftext := node.Data[ref.RefLocation.Start:ref.RefLocation.End]
	if _, ok := ctx.metas["format"]; ok {
		ctx.metas["index"] = ref.Issue
		link = createLink(com.Expand(ctx.metas["format"], ctx.metas), reftext, "issue")
	} else if ref.Owner == "" {
		link = createLink(util.URLJoin(setting.AppURL, ctx.metas["user"], ctx.metas["repo"], **reftype**, ref.Issue), reftext, "issue")
	} else {
		link = createLink(util.URLJoin(setting.AppURL, ref.Owner, ref.Name, **reftype**, ref.Issue), reftext, "issue")
	}

```

Reply:
>Besides, the point is to be backward compatible. If everybody likes #1234 and use Gitea for tracking issues, I don't see any problem keeping that syntax universal.

It will be backward compatible, `#1234` will still resolve to https://site.com/org/repo/issues/1234. If you click it and it actually is a PR you will be redirected to https://site.com/org/repo/pulls/1234 (as today). This will just be the opposite way, `!1234` is resolved to https://site.com/org/repo/pulls/1234, but you will be redirected if it is an issue.

Reply:
I see your point. But here's my reasoning:

* If for some reason some user decides to _always_ use `!`, the Gitea will render a `/pulls/` link every time. And in Gitea a PR is an issue but an issue is not necessarily a PR. IMHO it's more correct to generalize the link as `/issues/`.
* The `!` notation is conceptualized in the code as _local_ links, not _PR_ links, so rendering a PR link is not consistent.

I can change it, though.

Reply:
I see our opinions differ here on what `!` should represent. The difference in the end result will be quite small so it is not a very big deal. Personally I would think of `!` as pull request when I used it, even if I would only use Gitea internal issues/pull requests. I don't see why someone would use always `!` for both issues and pull requests, rather it would be a way to easily distinguish between issue and pull requests references when you read the text. Also I don't think there ever will be support for both internal and external issues on the same repository, so the _local_ reference doesn't make much sense to me.

Regarding references to issues in other repositories: I would say this is another argument for having `!` meaning pulls. Since the renderer can't/shouldn't check the database if a certain repository uses external issue tracker, it could be redirected when trying to access the issue. So if you refer to `org/repo#123` it would render to https://site.com/org/repo/pulls/123, while if you refer to `org/repo!123` it would render https://site.com/org/repo/issues/123. When you click this and the repo is set to use external issue tracker it can redirect to the external issue tracker instead.

Reply:
As my last paragraph above would resolve the combinations you crossed out, it actually is my strongest argument for why `!` should mean pulls and give a link with _pulls_ instead of _issues_.

Reply:
@davidsvantesson Done. 😄